### PR TITLE
Add basic teleportation and portals

### PR DESCRIPTION
**Pull request by metalgearsloth**
I'm in the discord if you want to message me there. Apologies for anything dumb in advance as I don't have much programming experience.

Shouldn't be too much more work to add beacons and gateways on top of this at some stage.

### CODE REVIEW

**Comment by Injazz** [comment]
image

File: Content.Client/GameObjects/EntitySystems/ItemTeleporterSystem.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+using Robust.Client.Interfaces.Input;
+using Robust.Client.Player;
+using Robust.Shared.GameObjects.Systems;
+using Robust.Shared.Input;
+using Robust.Shared.IoC;
+
+namespace Content.Client.GameObjects.EntitySystems
+{
+    [UsedImplicitly]
+    public class ItemTeleporterSystem : EntitySystem
```
Comment (PJB3005) [inline Content.Client/GameObjects/EntitySystems/ItemTeleporterSystem.cs]: Why can't all of this be handled on the server? This is quite a lot of fragile code.

File: Content.Server/GameObjects/Components/Movement/ServerPortalComponent.cs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            if (_connectingTeleporter != null &&
+                _connectingTeleporter.TryGetComponent<ServerPortalComponent>(out var otherPortal))
+            {
+                otherPortal.immuneEntities.Remove(entity);
+            }
+        }
+
+        private IEnumerable<IEntity> GetPortableEntities()
+        {
+            // TODO: Performance optimisation?
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Movement/ServerPortalComponent.cs]: As a performance optimization, you can cut out the list creation and return and replace `portableEntities.Add(entity);` with `yield return entity;`. No need to allocate a list then.

File: Content.Server/GameObjects/Components/Movement/ServerPortalComponent.cs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                if (entity.HasComponent<TeleportableComponent>())
+                {
+                    portableEntities.Add(entity);
+                }
+            }
+
+            return portableEntities;
+        }
+
+        public void OnUpdate()
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Movement/ServerPortalComponent.cs]: I believe there should be events on `ICollidable` that get invoked when the entity starts overlapping. Use those and you won't have to expensively check every tick like this.

Comment (metalgearsloth) [reply Content.Server/GameObjects/Components/Movement/ServerPortalComponent.cs]: I saw BumpedEntMsg under collidable so I could use that instead?

File: Content.Client/GameObjects/EntitySystems/ItemTeleporterSystem.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Client/GameObjects/EntitySystems/ItemTeleporterSystem.cs]: Yeah you're right. I ended up copying the ranged weapon system but that was dumb in hindsight as they're different use cases.

I'm trying to use IAttack but I can't seem to get it working.

Comment (PJB3005) [reply Content.Client/GameObjects/EntitySystems/ItemTeleporterSystem.cs]: You want `IAfterAttack` or `IAttack` instead. `IAttack` is used for when the mob is in combat mode (which isn't 100% implemented yet).

I only now realize the ridiculous naming here jeez.

File: Content.Server/GameObjects/Components/Movement/ServerPortalComponent.cs (hunk 3 above)
Comment (PJB3005) [reply Content.Server/GameObjects/Components/Movement/ServerPortalComponent.cs]: Looking at it appears there actually isn't any such event. In that case just slap a TODO on it and keep it like this. Acruid was rewriting physics anyways soo..

**Pull request merged**